crates/biome_module_graph/src/js_module_info/diagnostics.rs

```diff
@@ -0,0 +1,141 @@
+        match self {
+            Self::ExceededTypesLimit(d) => d.message(fmt),
+        }
+    }
+
+    fn description(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::ExceededTypesLimit(d) => d.description(fmt),
+        }
+    }
+
+    fn source(&self) -> Option<&dyn Diagnostic> {
+        match self {
+            Self::ExceededTypesLimit(d) => d.source(),
+        }
+    }
+
+    fn location(&self) -> Location<'_> {
+        match self {
+            Self::ExceededTypesLimit(d) => d.location(),
+        }
+    }
+
+    fn tags(&self) -> DiagnosticTags {
+        match self {
+            Self::ExceededTypesLimit(d) => d.tags(),
+        }
+    }
+
+    fn advices(&self, visitor: &mut dyn Visit) -> std::io::Result<()> {
+        match self {
+            Self::ExceededTypesLimit(d) => d.advices(visitor),
+        }
+    }
+
+    fn category(&self) -> Option<&'static Category> {
+        match self {
+            Self::ExceededTypesLimit(d) => d.category(),
+        }
+    }
+
+    fn verbose_advices(&self, visitor: &mut dyn Visit) -> std::io::Result<()> {
+        match self {
+            Self::ExceededTypesLimit(d) => d.verbose_advices(visitor),
+        }
+    }
+}
+
+impl JsModuleInfoDiagnostic {
+    pub fn exceeded_types_limit() -> Self {
+        Self::ExceededTypesLimit(ExceededTypesLimitDiagnostic)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ExceededTypesLimitDiagnostic;
+
+impl Diagnostic for ExceededTypesLimitDiagnostic {
+    fn severity(&self) -> Severity {
+        Severity::Warning
+    }
+
+    fn category(&self) -> Option<&'static Category> {
+        Some(category!("project"))
+    }
+
+    fn message(&self, fmt: &mut Formatter<'_>) -> std::io::Result<()> {
+        fmt.write_markup(markup! {
+            "Biome encountered an unusually large amount of types which exceeded the limit of "{MAX_NUM_TYPES}"."
+        })?;
+
+        fmt.write_str("\n\n")?;
+
+        fmt.write_markup(markup! {
+            "Either you are analyzing "<Emphasis>"very"</Emphasis>" large files (did you make sure to \
+        exclude your "<Emphasis>"build/"</Emphasis>" or "<Emphasis>"dist/"</Emphasis>" folder?), or you've encountered a bug in \
+        Biome."
+        })?;
+
+        fmt.write_str("\n\n")?;
+
+        fmt.write_markup(markup! {
+            "Please follow these instructions to discover if you are accidentally \
+        analyzing large files and what to do about them in "<Hyperlink href={"https://biomejs.dev/guides/investigate-slowness/"}>"the relative guide."</Hyperlink>
+        })?;
+
+        Ok(())
+    }
+
+    fn advices(&self, visitor: &mut dyn Visit) -> std::io::Result<()> {
+        visitor.record_log(LogCategory::Info, &markup!{
+            "In the meantime, you can ignore this file by adding its name or folder in the "<Emphasis>"files.experimentalScannerIgnores"</Emphasis>" option in your configuration file."
```

Comment:
Now that we bubble up the diagnostic. Might it be possible to add the path to the offending file to the diagnostic? If we can, we make it not only easier for users to diagnose their problems, but also for them to report offending files to us.

Reply:
> Might it be possible to add the path to the offending file to the diagnostic? 

Yes, we're adding the path in the `open_file` function, where we handle the `catch_unwind`

<img width="1370" height="408" alt="Screenshot 2025-08-20 at 15 05 24" src="https://github.com/user-attachments/assets/d71f68dd-9a15-4d84-8acb-40af424459d9" />